pkgs/stdenv/generic/setup.sh

```diff
@@ -199,6 +199,29 @@ isELF() {
     if [ "$magic" = $'\177ELF' ]; then return 0; else return 1; fi
 }
 
+# Return success if the specified file is a Mach-O object.
+isMachO() {
+    local fn="$1"
+    local fd
+    local magic
+    exec {fd}< "$fn"
+    read -r -n 4 -u "$fd" magic
```

Comment:
We can't use file, can we?

Reply:
`isELF` is doing it with `read` so I just followed.

I don't know how robust is:

```
$ file ~/.nix-profile/bin/nix | grep -q 'Mach-O'
```

There is [`libmagic`](https://man7.org/linux/man-pages/man3/libmagic.3.html) too, provided by `file`.

Reply:
> $ file ~/.nix-profile/bin/nix | grep -q 'Mach-O'

Doing this would be dirt cheap and would require zero maintenance.

Reply:
> Doing this would be dirt cheap and would require zero maintenance.

So are you going to contribute a PR?

Reply:
I can update the PR but how to do it is not my decision. I just ensured to do it the same way as `isELF`, which seems to work just fine, and use the proper magic numbers.

Here are the `isELF` usages:

```
$ rg --sort=path 'isELF '
pkgs/applications/misc/blender/default.nix
152:      isELF "$program" || continue

pkgs/applications/science/machine-learning/sc2-headless/default.nix
43:      isELF "$file" || continue

pkgs/build-support/add-opengl-runpath/setup-hook.sh
24:        if ! isELF "$file"; then continue; fi

pkgs/build-support/setup-hooks/audit-blas.sh
10:        if ! isELF "$i"; then continue; fi

pkgs/build-support/setup-hooks/audit-tmpdir.sh
22:        if isELF "$i"; then

pkgs/build-support/setup-hooks/auto-patchelf.sh
220:      isELF "$file" || continue

pkgs/build-support/setup-hooks/separate-debug-info.sh
19:        if ! isELF "$i"; then continue; fi

pkgs/development/compilers/cudatoolkit/common.nix
189:      if ! isELF "$i"; then continue; fi

pkgs/development/compilers/ghc/8.10.2-binary.nix
290:        if isELF "$p"; then
299:        if isELF "$p"; then

pkgs/development/compilers/ghc/8.6.5-binary.nix
138:      if isELF "$p"; then

pkgs/development/compilers/openjdk/bootstrap.nix
35:      isELF "$elf" || continue

pkgs/development/tools/misc/patchelf/setup-hook.sh
16:        if ! isELF "$i"; then continue; fi
```

Reply:
@smancill Sorry, that question was to @SuperSandro2000

Reply:
> Here are the `isELF` usages:

Yes, so the reason to have isELF in stdenv is for use inside the setup hooks. Same is the plan for isMachO.

Reply:
> So are you going to contribute a PR?

I don't really have the time for that.

Reply:
`file` is not part of stdenv, adding it there is not dirt cheap. You can find any PR for file and see that it's not an stdenv rebuild. Sending people down that path is not respectful of their time, would you agree @SuperSandro2000?

Reply:
Well, if it is not part of stdenv and this approach is alright. In most other scenarios this would be an unnecessary workaround. I don't know which tools are exactly part of stdenv but file seems to be unix-core enough that it could have been.